docs/solidity-by-example.rst

```diff
@@ -645,4 +645,489 @@ Safe Remote Purchase
+                v := byte(0, mload(add(sig, 96)))
+            }
+            
+            return (v, r, s);
+        }
+        
+        function recoverSigner(bytes32 message, bytes sig)
+            internal
+            pure
+            returns (address)
+        {
+            (uint8 v, bytes32 r, bytes32 s) = splitSignature(sig);
+
+            return ecrecover(message, v, r, s);
+        }
+
+        /// builds a prefixed hash to mimic the behavior of eth_sign.
+        function prefixed(bytes32 hash) internal pure returns (bytes32) {
+            return keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", hash));
+        }
+    }
+
+
+Note: The function ``splitSignature`` is very simple and does not use all security checks.
+A real implementation should use a more rigorously tested library, such as
+openzepplin's `version <https://github.com/OpenZeppelin/openzeppelin-solidity/blob/master/contracts/ECRecovery.sol>`_ of this code.
+
+  
+
+Verifying Payments
+------------------
+
+Unlike in our previous chapter, messages in a payment channel aren't
+redeemed right away. The recipient keeps track of the latest message and
+redeems it when it's time to close the payment channel. This means it's
+critical that the recipient perform their own verification of each message.
+Otherwise there is no guarantee that the recipient will be able to get paid
+in the end.
+
+The recipient should verify each message using the following process:
+
+    1. Verify that the contact address in the message matches the payment channel.
+    2. Verify that the new total is the expected amount.
+    3. Verify that the new total does not exceed the amount of Ether escrowed.
+    4. Verify that the signature is valid and comes from the payment channel sender.
+    
+We'll use the `ethereumjs-util <https://github.com/ethereumjs/ethereumjs-util>`_
+library to write this verifications. The final step can be done a number of ways,
+but if it's being done in **JavaScript**.
+The following code borrows the `constructMessage` function from the signing **JavaScript code**
+above:
+
+::
+
+    // this mimics the prefixing behavior of the eth_sign JSON-RPC method.
+    function prefixed(hash) {
+        return ethereumjs.ABI.soliditySHA3(
+            ["string", "bytes32"],
+            ["\x19Ethereum Signed Message:\n32", hash]
+        );
+    }
+    
+    function recoverSigner(message, signature) {
+        var split = ethereumjs.Util.fromRpcSig(signature);
+        var publicKey = ethereumjs.Util.ecrecover(message, split.v, split.r, split.s);
+        var signer = ethereumjs.Util.pubToAddress(publicKey).toString("hex");
+        return signer;
+    }
+    
+    function isValidSignature(contractAddress, amount, signature, expectedSigner) {
+        var message = prefixed(constructPaymentMessage(contractAddress, amount));
+        var signer = recoverSigner(message, signature);
+        return signer.toLowerCase() ==
```

Comment:
I think this should be `===` as part of latest JS standards, but in this case it sohuldn't matter too much.